### PR TITLE
fix seqNum concurrency issue in Queue.write

### PR DESCRIPTION
**Pull request by colinsurprenant**
The `Queue.write()` method need a new `seqNum` and this operation was done outside the synchronized section so concurrent writes could result in being assigned invalid `seqNum` since it is not volatile. 

This solves #6829 using @jordansissel provided reproduction spec (https://gist.github.com/jordansissel/518667ab4bd2df9d41244427917c9160)

This is an "early" PR and would like to review all reported problems in #6829 to make sure it fixes all. Also I would like to include a spec/test for this, should we include @jordansissel one as it is quite long to run?

### CODE REVIEW

**Comment by talevy** [comment]
> should we include @jordansissel one as it is quite long to run?

maybe we can introduce a new build-test task in gradle that runs that type of test separately?

**Comment by colinsurprenant** [comment]
or add Jordan's spec as an integration test? I will create a Java test for this, we should be able to reproduce the concurrency bug in a simple test.

**Comment by colinsurprenant** [comment]
I wrote a Java concurrent writes test and this PR does indeed fix the seqNum problem but this test is showing that some fully acked tail pages are not purged occasionally. All the lingering tail pages are fully acked but still in the `tailPages` list for some reason. 

Working on it.

**Comment by colinsurprenant** [comment]
I found the bug that created lingering pages.
The head page beheading did not verify if the head page was already fully acked before transforming it into a tail page in which case this transformation is not necessary. Because there wasn't this fully acked check, a fully acked tail page was created but not purged.

this last commit includes the fix and a new test for this specific case. Also, my previous test for `seqNum` is now always passing.

**Comment by colinsurprenant** [comment]
@jordansissel @guyboertje can you rerun your tests against this PR and confirm the fix?

**Comment by guyboertje** [comment]
My specs are still failing See below. The queue reopen failure is gone but there are still orphaned or lingering pages with no checkpoint.

If I understand correctly - a new headpage is created if there is no space, but the old head page will only be beheaded and put in the tailPages collection if it still has unacked events. All good. However a fully acked headpage cannot be purged, only tailPages can, we need to create a tailPage using the headPage and purge it when its fully acked.

I was able to get this bit of code to pass my specs (but only by standing on the shoulders of Giants).
```java
            boolean wasEmpty = (firstUnreadPage() == null);
            // create a new head page if the current does not have sufficient space left for data to be written
            if (! this.headPage.hasSpace(data.length)) {
                int newHeadPageNum = this.headPage.pageNum + 1;
                TailPage tailPage;
                if (this.headPage.isFullyAcked()) {
                    tailPage = new TailPage(this.headPage);
                    // purge the old headPage because its full and fully acked
                    // there is no checkpoint file to purge
                    tailPage.purge();
                } else {
                    // beheading includes checkpoint+fsync if required
                    tailPage = this.headPage.behead();
                    this.tailPages.add(tailPage);
                    if (!tailPage.isFullyRead()) {
                        this.unreadTailPages.add(tailPage);
                    }
                }

                // create new head page
                newCheckpointedHeadpage(newHeadPageNum);
            }
``` 

```
$ rake test:core-single-file[/Users/guy/elastic/logstash/logstash-core/spec/logstash/util/threaded_queue_spec.rb]
........F.....FF......FF....

Failures:

  1) LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work ON > larger event size < behaves like a well behaved queue writes, reads, closes and reopens
     Failure/Error: expect(output_strings).to eq([])

       expected: []
            got: ["File count after close mismatch expected: 2 got: 18", "checkpoint.157", "checkpoint.160", "checkpoint.173", "checkpoint.179", "checkpoint.197", "checkpoint.203", "checkpoint.248", "checkpoint.329", "checkpoint.349", "checkpoint.354", "checkpoint.393", "checkpoint.400", "checkpoint.41", "checkpoint.425", "checkpoint.445", "checkpoint.83", "checkpoint.head", "page.476", "{:page_capacity_multiplier=>20, :page_capacity=>1048576, :queue_checkpoint_acks=>1024, :queue_checkpoint_writes=>1024, :queue_checkpoint_interval=>1000, :queue_capacity_multiplier=>128, :queue_capacity=>134217728, :writers=>3, :expected_count=>60000, :items=>20000, :string_size=>8192, :batch_size=>500, :batch_wait=>1000, :readers=>3, :simulate_work=>true, :counts=>[20084, 18762, 21154, 0, 0, 0, 0, 0]}"]

       (compared using ==)
     Shared Example Group: "a well behaved queue" called from ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:193
     # ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:113:in `(root)'

  2) LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > small queue size limit < behaves like a well behaved queue writes, reads, closes and reopens
     Failure/Error: expect(output_strings).to eq([])

       expected: []
            got: ["File count after close mismatch expected: 2 got: 3", "checkpoint.head", "page.19", "page.3", "{:page_capacity_multiplier=>20, :page_capacity=>1048576, :queue_checkpoint_acks=>1024, :queue_checkpoint_writes=>1024, :queue_checkpoint_interval=>1000, :queue_capacity_multiplier=>10, :queue_capacity=>10485760, :writers=>3, :expected_count=>60000, :items=>20000, :string_size=>256, :batch_size=>500, :batch_wait=>1000, :readers=>3, :simulate_work=>false, :counts=>[20523, 19396, 20081, 0, 0, 0, 0, 0]}"]

       (compared using ==)
     Shared Example Group: "a well behaved queue" called from ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:274
     # ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:113:in `(root)'

  3) LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > less writers than readers < behaves like a well behaved queue writes, reads, closes and reopens
     Failure/Error: expect(output_strings).to eq([])

       expected: []
            got: ["File count after close mismatch expected: 2 got: 3", "checkpoint.head", "page.19", "page.9", "{:page_capacity_multiplier=>20, :page_capacity=>1048576, :queue_checkpoint_acks=>1024, :queue_checkpoint_writes=>1024, :queue_checkpoint_interval=>1000, :queue_capacity_multiplier=>128, :queue_capacity=>134217728, :writers=>2, :expected_count=>60000, :items=>30000, :string_size=>256, :batch_size=>500, :batch_wait=>1000, :readers=>4, :simulate_work=>false, :counts=>[15049, 14861, 15033, 15057, 0, 0, 0, 0]}"]

       (compared using ==)
     Shared Example Group: "a well behaved queue" called from ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:219
     # ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:113:in `(root)'

  4) LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > larger event size < behaves like a well behaved queue writes, reads, closes and reopens
     Failure/Error: expect(output_strings).to eq([])

       expected: []
            got: ["File count after close mismatch expected: 2 got: 10", "checkpoint.head", "page.141", "page.191", "page.2", "page.217", "page.363", "page.391", "page.427", "page.455", "page.476", "{:page_capacity_multiplier=>20, :page_capacity=>1048576, :queue_checkpoint_acks=>1024, :queue_checkpoint_writes=>1024, :queue_checkpoint_interval=>1000, :queue_capacity_multiplier=>128, :queue_capacity=>134217728, :writers=>3, :expected_count=>60000, :items=>20000, :string_size=>8192, :batch_size=>500, :batch_wait=>1000, :readers=>3, :simulate_work=>false, :counts=>[20121, 19710, 20169, 0, 0, 0, 0, 0]}"]

       (compared using ==)
     Shared Example Group: "a well behaved queue" called from ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:269
     # ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:113:in `(root)'

  5) LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > larger checkpoint acks < behaves like a well behaved queue writes, reads, closes and reopens
     Failure/Error: expect(output_strings).to eq([])

       expected: []
            got: ["File count after close mismatch expected: 2 got: 3", "checkpoint.head", "page.19", "page.9", "{:page_capacity_multiplier=>20, :page_capacity=>1048576, :queue_checkpoint_acks=>3000, :queue_checkpoint_writes=>1024, :queue_checkpoint_interval=>1000, :queue_capacity_multiplier=>128, :queue_capacity=>134217728, :writers=>3, :expected_count=>60000, :items=>20000, :string_size=>256, :batch_size=>500, :batch_wait=>1000, :readers=>3, :simulate_work=>false, :counts=>[20408, 19607, 19985, 0, 0, 0, 0, 0]}"]

       (compared using ==)
     Shared Example Group: "a well behaved queue" called from ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:224
     # ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:113:in `(root)'

Finished in 2 minutes 28.8 seconds (files took 1.03 seconds to load)
28 examples, 5 failures

Failed examples:

rspec ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:72 # LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work ON > larger event size < behaves like a well behaved queue writes, reads, closes and reopens
rspec ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:72 # LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > small queue size limit < behaves like a well behaved queue writes, reads, closes and reopens
rspec ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:72 # LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > less writers than readers < behaves like a well behaved queue writes, reads, closes and reopens
rspec ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:72 # LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > larger event size < behaves like a well behaved queue writes, reads, closes and reopens
rspec ./logstash-core/spec/logstash/util/threaded_queue_spec.rb:72 # LogStash::Util::WrappedAckedQueue with multiple writers with simulate_work OFF > larger checkpoint acks < behaves like a well behaved queue writes, reads, closes and reopens

Randomized with seed 24307

```

**Comment by guyboertje** [comment]
BTW, looking at the failures in my comment above. In my spec, in general, the readers are as fast as the writers.
1) when the queue size is small the writers will block on queue full allowing the readers to be on the headPage most of the time.
2) when there is less writers than readers, the readers can keep up with the writers on the headPage.
3) when the event size is large, the writers spend much time building + writing individual events, the writers are slower than the readers and they can keep up.
4) when the `checkpoint_acks` are large, dunno.
This explains why in the wild we have not seen this very often, the readers are normally much slower than the writers and read + acking is being done on the tailPages.

**Comment by colinsurprenant** [comment]
@guy ah yes, of course - the physical pages are still there - I was working on the memory model and forgot the files - your suggested fix is good. commit incoming!

**Comment by colinsurprenant** [comment]
ok, pushed the fix to correctly remove preserved checkpoints. 

I also corrected the spec to fix the concurrency issue with the `counts` array.  https://gist.github.com/colinsurprenant/b3ca535857a0e5d4ab48580fdfd653fe

It now passes systematically on my end! 
I believe we are now golden. 
Really good job on the new spec @guyboertje @jordansissel  it did help fix these issues.

**Comment by jordansissel** [comment]
I will review this tomorrow <3

File: logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java [hunk 1]
```diff
@@ -638,4 +639,89 @@ public void testAckedCount() throws IOException {
...
+        assertThat(q.isFullyAcked(), is(true));
+
+        // write extra element to trigger beheading
+        q.write(element);
+
+        // since head page was fully acked it should not have created a new tail page
+
+        assertThat(q.getTailPages().isEmpty(), is(true));
+        assertThat(q.getHeadPage().getPageNum(), is(equalTo(1)));
+        assertThat(q.firstUnackedPageNum(), is(equalTo(1)));
```
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java]: Should we also assert that `q.isFullyAcked()` is false, or do you think that is out of scope for the test?

**Comment by jordansissel** [comment]
Marking LGTM because I trust the tests and that my questions are mostly style or curiosities, not blockers.

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java [hunk 2]
```diff
@@ -314,18 +321,32 @@ public long write(Queueable element) throws IOException {
...
+                    // there is no checkpoint file to purge since just creating a new TailPage from a HeadPage does
+                    // not trigger a checkpoint creation in itself
+                    TailPage tailPage = new TailPage(this.headPage);
+                    tailPage.purge();
+                } else {
+                    // beheading includes checkpoint+fsync if required
+                    TailPage tailPage = this.headPage.behead();
+
+                    this.tailPages.add(tailPage);
+                    if (! tailPage.isFullyRead()) {
```
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: A page could be not fully acked but be fully read, where the read elements are awaiting to be acked.

File: logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java [hunk 3]
```diff
@@ -638,4 +639,89 @@ public void testAckedCount() throws IOException {
...
+            b.close();
+        }
+
+        for (Future<Integer> future : futures) {
+            int result = future.get();
+            assertThat(result, is(equalTo(ELEMENT_COUNT)));
+        }
+
+        assertThat(q.getTailPages().isEmpty(), is(true));
+        assertThat(q.isFullyAcked(), is(true)); // queue should still be full
```
Comment (colinsurprenant) [inline logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java]: yes, copy-pasta error. will update.

File: logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java (hunk 1 above)
Comment (colinsurprenant) [inline logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java]: since we do assert that `assertThat(q.isFullyAcked(), is(true));` above, it makes sense to also assert he opposite since we changed that state. will add.

**Comment by colinsurprenant** [comment]
@jordansissel thanks! addressed your comments.

**Comment by jordansissel** [comment]
@colinsurprenant LGTM :)

**Comment by colinsurprenant** [comment]
Thanks @jordansissel @guyboertje @talevy - merging in 5.3.1 and 5.4

**Comment by colinsurprenant** [comment]
rebased & squashed, merged in master, 5.x, 5.4, 5.3
